core/dbt/parser/schemas.py

```diff
@@ -737,6 +746,50 @@ def add_source_definitions(self, source: UnparsedSourceDefinition) -> None:
             )
             self.manifest.add_source(self.yaml.file, source_def)
 
+class DynamicSourceParser(SourceParser):
+    def _target_from_dict(self, cls: Type[T], data: Dict[str, Any]) -> T:
+        return super()._target_from_dict(UnparsedDynamicSourceDefinition, data)
+
+    def add_source_definitions(self, source: UnparsedDynamicSourceDefinition) -> None:
+        original_file_path = self.yaml.path.original_file_path
+        fqn_path = self.yaml.path.relative_path
+        exclude_pattern = re.compile(source.tables_exclude_pattern)
+        include_pattern = re.compile(source.tables_include_pattern)
+        
+        tables = []
+        adapter = get_adapter(self.root_project)
+        with adapter.connection_named("dynamic_sources"):
+            schema_relation = BaseRelation.create(
+                # database = self.root_project.default_database,
+                schema = source.schema
+            )
+            relations = adapter.list_relations_without_caching(schema_relation)
```

Comment:
Hard rule in dbt: We shouldn't be making database connections at parse time. I said more about this over in https://github.com/dbt-labs/dbt-core/issues/6067#issuecomment-1296516720:

> There's a hard line we draw in dbt today, and don't wish to soon erase: dbt must know the full DAG before it connects to the database. This is a useful abstraction and simplification for us. We can keep "control plane" code (DAG definitions) clearly separate from "data plane" code (tabular queries / dataframe manipulations). Every dbt resource maps to exactly one database object (or zero, in the case of ephemeral models, analyses, etc).

> For that reason, it's essential that we don't require dbt to make a database connection — even if it's just a caching query — to enumerate the set of available sources.